dm/docs/RFCS/20220721_continuous_data_validation.md

```diff
@@ -0,0 +1,64 @@
+  - there is no operation to change primary key or non-null unique key
+  - there is no column order change operation
+  - there is no drop column operation
+  - table in downstream which needs validation cannot be dropped
+- do not support validation for tasks which enable extend-column
+- do not support validation for tasks which enable event filtering by expression
+- TiDB implements floating point data types differently with MySQL, we take it as equal if its absolute difference < 10^-6
+- do not support validation for JSON and binary data types
+
+## Detailed Design
+
+### Life cycle of validator
+
+Validator can be enabled together with the task or enabled on the fly, the task which enables validator should has a incremental migration unit, i.e. syncer. Validator will be deleted when the task is deleted. Enabled validator will be in `running` state, it can be stopped manually or stopped when met error that cannot be recovered automatically, then it's in `stopped` state. Validator in `stopped` state can be started again to return to `running` state.
+
+### Validation process
+
+1. Validator in `running` state will pull binlog from upstream and get `row change`s.
+    - validator will only validate `row change` which has already migrated by syncer.
+2. After routing and filtering using the same rule as in syncer, `row change` is dispatched to validator `worker`. `row change` of same table and primary key will be dispatched to the same `worker`.
+3. `worker` merges `row change` by table and primary key, we will use the last `row change` meet since last change override previous one. Then put it into `pending row changes`.
+4. After accumulates enough `row change`s or after some interval, `worker` queries downstream for data using table and primary key info in those `row change`s, then compare with data in `row change`s
+    - for `insert`/`update` type of row change, we validate them differently by validation mode
+      - in `full` validation mode, we compare them column by column
+      - in `fast` validation mode, we only check it's existence
+    - for `delete` type of `row change`, downstream should not contains that data.
+5. For `row change`s which are validated successfully, worker will remove it from `pending row changes`, others which fail to validate will be marked as `failed row change` and be validated again after some interval.
+6. If a `failed row change` keeps failing to validate for enough time(`delay time`), we mark it as `error row change` and save it into meta database.
+
+### False positive
+
+`error row change` produced in validation process might not be data which is incorrectly migrated, there are cases where `row change` is marked as falsely. Suppose some row which keeps changing on upstream for a time period > `delay time`. If it's marked as `failed row change` since the first time it changes, validator may mark it as `error row change` falsely. In real world scenarios, it's not common.
```

Comment:
```suggestion
`error row change` produced in validation process might not be data which is incorrectly migrated, there are cases where `row change` is marked error falsely. Suppose some rows keep changing on upstream for a time period > `delay time`. If it's marked as `failed row change` since the first time it changes, validator may mark it as `error row change` falsely. In real world scenarios, it's not common.
```